end-to-end-tests/src/helpers/ctx.rs

```diff
@@ -84,32 +86,168 @@ pub fn nexus_addr() -> SocketAddr {
+        || async {
+            let addr = resolver
+                .lookup_ip(&dns_name)
+                .await
+                .map_err(|e| match e.kind() {
+                    ResolveErrorKind::NoRecordsFound { .. }
+                    | ResolveErrorKind::Timeout => CondCheckError::NotYet,
+                    _ => CondCheckError::Failed(anyhow::Error::new(e).context(
+                        format!("resolving {:?} from {}", dns_name, dns_addr),
+                    )),
+                })?
                 .iter()
-                .flat_map(|range| range.iter())
-                .skip(1)
                 .next()
-            {
-                return (addr, 80).into();
-            }
-        }
-    }
-
-    ([192, 168, 1, 21], 80).into()
+                .ok_or(CondCheckError::NotYet)?;
+            Ok(SocketAddr::from((addr, 80)))
+        },
+        &Duration::from_secs(1),
+        &Duration::from_secs(300),
+    )
+    .await
+    .unwrap_or_else(|e| panic!("failed to get Nexus addr: {}", e))
 }
 
-fn get_base_url() -> String {
+async fn get_base_url() -> String {
     // Check $OXIDE_HOST first.
     if let Ok(host) = std::env::var("OXIDE_HOST") {
         return host;
     }
 
-    format!("http://{}", nexus_addr())
+    format!("http://{}", nexus_addr().await)
+}
+
+async fn build_authenticated_client() -> Result<oxide_client::Client> {
+    let config = rss_config()?;
+    let base_url = get_base_url().await;
+    let silo_name = config.recovery_silo.silo_name.as_str();
+    let username: oxide_client::types::UserId =
+        config.recovery_silo.user_name.as_str().parse().map_err(|s| {
+            anyhow!("parsing configured recovery user name: {:?}", s)
+        })?;
+    // See the comment in the config file.
+    let password: oxide_client::types::Password = "oxide".parse().unwrap();
+    let login_request_body =
+        serde_json::to_string(&UsernamePasswordCredentials {
+            username: username,
+            password: password,
+        })
+        .context("serializing login request body")?;
+
+    // Do not have reqwest follow redirects.  That's because our login response
+    // includes both a redirect and the session cookie header.  If reqwest
+    // follows the redirect, we won't have a chance to get the cookie.
+    let reqwest_login_client = reqwest::ClientBuilder::new()
+        .connect_timeout(Duration::from_secs(15))
+        .redirect(reqwest::redirect::Policy::none())
+        .timeout(Duration::from_secs(60))
+        .build()?;
+    let login_url = format!("{}/login/{}/local", base_url, silo_name);
+
+    // By the time we get here, we generally would have successfully resolved
+    // Nexus's external IP address from the external DNS server.  So we'd
+    // expect Nexus to be up.  But that's not necessarily true: external DNS can
+    // be set up during rack initialization, before Nexus has opened its
+    // external listening socket.  This is arguably a bug, advertising a service
```

Comment:
Should we create an issue to track that? I know the rack/service initialization process could still use some polish.

Reply:
You mean to track the behavior that Nexus publishes its addresses to DNS before it's opened its listening socket?  I'm  not clear that's a bug but I'm happy to file an issue to document it and/or have that discussion.

Reply:
If the expectation is that we should only be adding the DNS records once the service is up then seems we should have an issue tracking that